pkg/sentry/fsimpl/fuse/connection.go

```diff
@@ -0,0 +1,241 @@
+	// Create the writeBuf for the header to be stored in.
+	hdrLen := uint32((*linux.FUSEHeaderOut)(nil).SizeBytes())
+	fuseFD.writeBuf = make([]byte, hdrLen)
+	fuseFD.completions = make(map[linux.FUSEOpID]*futureResponse)
+	fuseFD.requestKind = make(map[linux.FUSEOpID]linux.FUSEOpcode)
+	fuseFD.emptyQueueCh = make(chan struct{}, maxInFlightRequests)
+	fuseFD.fullQueueCh = make(chan struct{}, maxInFlightRequests)
+
+	// This is emulating the behaviour of a counting semaphore. Is there a better
+	// way to do this?
+	for i := uint64(0); i < maxInFlightRequests; i++ {
+		fuseFD.fullQueueCh <- struct{}{}
+	}
+
+	fuseFD.writeCursor = 0
+	fuseFD.readCursor = 0
+
+	return &Connection{
+		fd: fuseFD,
+	}, nil
+}
+
+// NewRequest creates a new request that can be sent to the FUSE server.
+func (conn *Connection) NewRequest(creds *auth.Credentials, pid uint32, ino uint64, opcode linux.FUSEOpcode, payload marshal.Marshallable) (*Request, error) {
+	conn.fd.mu.Lock()
+	defer conn.fd.mu.Unlock()
+	conn.fd.nextOpID += linux.FUSEOpID(ReqIDStep)
+
+	hdrLen := (*linux.FUSEHeaderIn)(nil).SizeBytes()
+	hdr := linux.FUSEHeaderIn{
+		Len:    uint32(hdrLen + payload.SizeBytes()),
+		Opcode: opcode,
+		Unique: conn.fd.nextOpID,
+		NodeID: ino,
+		UID:    uint32(creds.EffectiveKUID),
+		GID:    uint32(creds.EffectiveKGID),
+		PID:    pid,
+	}
+
+	buf := make([]byte, hdr.Len)
+	hdr.MarshalUnsafe(buf[:hdrLen])
+	payload.MarshalUnsafe(buf[hdrLen:])
+
+	return &Request{
+		id:   hdr.Unique,
+		hdr:  &hdr,
+		data: buf,
+	}, nil
+}
+
+// Call makes a request to the server and blocks the invoking task until a
+// server responds with a response.
+// NOTE: If no task is provided then the Call will simply enqueue the request
+// and return a nil response. No blocking will happen in this case. Instead,
+// this is used to signify that the processing of this request will happen by
+// the kernel.Task that writes the response. See FUSE_INIT for such an
+// invocation.
+func (conn *Connection) Call(t *kernel.Task, r *Request) (*Response, error) {
+	fut, err := conn.callFuture(t, r)
+	if err != nil {
+		return nil, err
+	}
+
+	return fut.resolve(t)
+}
+
+// callFuture makes a request to the server and returns a future response.
+// Call resolve() when the response needs to be fulfilled.
+func (conn *Connection) callFuture(t *kernel.Task, r *Request) (*futureResponse, error) {
+	// Is the queue full?
+	if conn.fd.numInFlightRequests == conn.fd.fs.opts.maxInflightRequests {
+		// Can't add a new request into the queue until space is cleared up.
+		if t == nil {
+			// Since there is no task that is waiting. We must error out.
+			return nil, errors.New("FUSE request queue full")
+		}
+	}
+
+	// Consider possible starvation here if the queue is continuously full.
+	// Will go channels respect FIFO order when unblocking threads?
+	if err := t.Block(conn.fd.fullQueueCh); err != nil {
+		log.Warningf("Connection.Call: couldn't wait on request queue: %v", err)
+		return nil, syserror.EBUSY
```

Comment:
If a task received a signal, we need to return syserror.ErrInterrupted.

I think you need to return err here:
  return nil, err;